temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java

```diff
@@ -0,0 +1,383 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.failure;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.TypeToken;
+import io.temporal.api.common.v1.ActivityType;
+import io.temporal.api.common.v1.Payload;
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.api.common.v1.WorkflowType;
+import io.temporal.api.failure.v1.ActivityFailureInfo;
+import io.temporal.api.failure.v1.ApplicationFailureInfo;
+import io.temporal.api.failure.v1.CanceledFailureInfo;
+import io.temporal.api.failure.v1.ChildWorkflowExecutionFailureInfo;
+import io.temporal.api.failure.v1.Failure;
+import io.temporal.api.failure.v1.ResetWorkflowFailureInfo;
+import io.temporal.api.failure.v1.ServerFailureInfo;
+import io.temporal.api.failure.v1.TerminatedFailureInfo;
+import io.temporal.api.failure.v1.TimeoutFailureInfo;
+import io.temporal.client.ActivityCanceledException;
+import io.temporal.common.converter.DataConverter;
+import io.temporal.common.converter.DefaultDataConverter;
+import io.temporal.common.converter.EncodedValues;
+import io.temporal.serviceclient.CheckedExceptionWrapper;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultFailureConverter implements FailureConverter {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultFailureConverter.class);
+
+  public static final String JAVA_SDK = "JavaSDK";
+
+  private static final String ENCODED_FAILURE_MESSAGE = "Encoded failure";
+  private static final String STACK_TRACE_KEY = "stack_trace";
+  private static final String MESSAGE_KEY = "message";
+
+  private static final Type HASH_MAP_STRING_STRING_TYPE =
+      new TypeToken<HashMap<String, String>>() {}.getType();
+
+  /**
+   * Stop emitting stack trace after this line. Makes serialized stack traces more readable and
+   * compact as it omits most of framework level code.
+   */
+  private static final ImmutableSet<String> CUTOFF_METHOD_NAMES =
+      ImmutableSet.of(
+          "io.temporal.internal.worker.POJOActivityImplementationFactory$POJOActivityImplementation.execute",
+          "io.temporal.internal.sync.POJOWorkflowTaskHandler$POJOWorkflowImplementation.execute");
+
+  /** Used to parse a stack trace line. */
+  private static final String TRACE_ELEMENT_REGEXP =
+      "((?<className>.*)\\.(?<methodName>.*))\\(((?<fileName>.*?)(:(?<lineNumber>\\d+))?)\\)";
```

Comment:
IMO you don't need all these single-reference static fields, but especially this one. Just inline it in the field below.

Reply:
I think giving cryptic strings a name and a separate place that makes it easier to write a javadoc next to it is a good practice for documentation purposes. I would leave it like this.

Reply:
This is not that case of cryptic string used elsewhere. This is just a string that's a single parameter to a regex creation below. If Java had auto-regex strings this wouldn't be a separate field. Not sure it makes sense to have every static final regex pattern be two fields.

```java
  /** Used to parse a stack trace line. */
  private static final String TRACE_ELEMENT_REGEXP =
      "((?<className>.*)\\.(?<methodName>.*))\\(((?<fileName>.*?)(:(?<lineNumber>\\d+))?)\\)";

  private static final Pattern TRACE_ELEMENT_PATTERN = Pattern.compile(TRACE_ELEMENT_REGEXP);
```

Provides 0 value over:

```java
  /** Used to parse a stack trace line. */
  private static final Pattern TRACE_ELEMENT_PATTERN = Pattern.compile(
      "((?<className>.*)\\.(?<methodName>.*))\\(((?<fileName>.*?)(:(?<lineNumber>\\d+))?)\\)");
```